entity-framework/core/querying/client-eval.md

```diff
@@ -1,69 +1,62 @@
-    }
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientProjection)]
 
-    return url;
-}
-```
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientMethod)]
 
-## Client evaluation performance issues
+## Unsupported client evaluation
 
-While client evaluation can be very useful, in some instances it can result in poor performance. Consider the following query, where the helper method is now used in a filter. Because this can't be performed in the database, all the data is pulled into memory and then the filter is applied on the client. Depending on the amount of data, and how much of that data is filtered out, this could result in poor performance.
+While client evaluation can be very useful, in some instances it can result in poor performance. Consider the following query, in which the helper method is now used in a where filter. Because this can't be performed in the database, all the data needs to be pulled into memory and then the filter needs to be applied on the client. Depending on the amount of data on the server, and how much of that data is filtered out, this could result in poor performance. Hence Entity Framework Core blocks such client evaluation and throws a runtime exception.
 
-<!-- [!code-csharp[Main](samples/core/Querying/ClientEval/Sample.cs)] -->
-``` csharp
-var blogs = context.Blogs
-    .Where(blog => StandardizeUrl(blog.Url).Contains("dotnet"))
-    .ToList();
-```
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ClientWhere)]
+
+## Explicit client evaluation
+
+In certain instances user may need to force into client evaluation explicitly like following cases
+
+- Data is small so that doing evaluation on client does not cause huge performance penalty.
+- LINQ operator being used has no server side translation.
+
+In such cases, user can explicitly opt into client evaluation by calling methods like `AsEnumerable` (or `AsAsyncEnumerable` for async).
+
+[!code-csharp[Main](../../../samples/core/Querying/ClientEval/Sample.cs#ExplicitClientEval)]
+
+## Potential memory leak in client evaluation
+
+Since query translation and compilation is expensive, by default, EF Core caches compiled query plan. With client evaluation allowed in the top projection, the cached delegate can use client code. In most cases, EF Core generates parameters for client parts of the tree so that their values can be replaced when reusing the query plan. But with certain constants in expression tree, it cannot be converted to parameters. If the client code contains such constants then those objects cannot be garbage collected. If such object contains DbContext or other services inside it then it could cause memory usage of the app to grow over time. In order to avoid potential memory leak, EF Core throws exception whenever it encounter constants of type which cannot be mapped in current database providers. Common causes and their solutions are as follows:
```

Comment:
```suggestion
Since query translation and compilation is expensive, by default, EF Core caches the compiled query plan. With client evaluation allowed in the top projection, the cached delegate can use client code. In most cases, EF Core generates parameters for client parts of the tree so that their values can be replaced when reusing the query plan. But certain constants in expression tree cannot be converted to parameters. If the client code contains such constants then those objects cannot be garbage collected as they would still be referenced from the cached query plan. If such object contains DbContext or other services inside it then it could cause memory usage of the app to grow over time. In order to avoid potential memory leak, EF Core throws an exception whenever it encounters constants of type which cannot be mapped using the current database provider. Common causes and their solutions are as follows:
```